src/jit/earlyprop.cpp

```diff
@@ -296,6 +296,33 @@ bool Compiler::optEarlyPropRewriteTree(GenTreePtr tree)
                 // array length argument, but the type of GT_ARR_LENGTH is always INT32.
                 return false;
             }
+
+            // When replacing GT_ARR_LENGTH nodes with constants we can end up with GT_ARR_BOUNDS_CHECK
+            // nodes that have constant operands and thus can be trivially proved to be useless. It's
+            // better to remove these range checks here, otherwise they'll pass through assertion prop
+            // (creating useless (c1 < c2)-like assertions) and reach RangeCheck where they are finally
+            // removed. Common patterns like new[] { 1, 2, 3 } benefit from this.
+
+            if ((tree->gtNext != nullptr) && tree->gtNext->OperIs(GT_ARR_BOUNDS_CHECK))
+            {
+                GenTreeBoundsChk* check = tree->gtNext->AsBoundsChk();
+
+                if (check->gtIndex->IsCnsIntOrI() && (static_cast<uint32_t>(check->gtIndex->AsIntCon()->IconValue()) <
+                                                      static_cast<uint32_t>(actualVal->AsIntCon()->IconValue())))
+                {
```

Comment:
These static_casts are pretty ugly, but I guess they are necessary, if you want to use a single unsigned compare operation. Instead you could use two compares here, since it is a compile time operation.

Reply:
I think casting to a 32 bit integer might still be necessary even if 2 compares are used. Need to check, I think I once fixed a bug that was caused by a `TYP_INT` icon having its upper 32 bit messed up somehow.

Reply:
> I think I once fixed a bug that was caused by a TYP_INT icon having its upper 32 bit messed up somehow.

I was thinking about #9336. Perhaps we should have a `template<T> T IconValue()` that returns the value converted to `T` and at the same time it ensures that the actual node type (`TYP_INT` or `TYP_LONG`) is compatible with the requested `T`. Or, to keep things simple just `IconValue32()`, similar to `GetConstantInt32` from VN.